### PR TITLE
box: add hostname to box.info

### LINKED ISSUE #8605: Add box.info.hostname
The hostname is a useful piece of information to supplement various instance state reports.

It looks convenient to have the hostname information together with other instance related details -- in `box.info`.

Note 1: Should be accessible as `box.info.hostname` and as `box.info().hostname` (for iproto calls) -- like other `box.info` fields.

Note 2: As regarding very long hostnames -- I would report `nil` (meaning 'cannot determine'), not a trimmed hostname. `gethostname()` on Linux reports `ENAMETOOLONG` in the case, but POSIX.1 says nothing about errors. Need to check macOS and FreeBSD behavior in this regard.

Note 3: The hostname may changes, so we should obtain it when requested, not cache at startup.

<details>
<summary>moonlibs/kit</summary>

https://github.com/moonlibs/kit/blob/7b46d4a8aba14944f1c924f16938dc2c8200bd8b/kit/1.lua#L50-L68
</details>

<details>
<summary>My snippet</summary>

```lua
-- utils/hostname.lua
local ffi = require('ffi')
local errno = require('errno')

-- Ignore errors to make the code reloadable.
pcall(ffi.cdef, 'int gethostname(char *name, size_t len);')

local function get()
    local buf_size = 256
    local buf = ffi.new('char[?]', buf_size)
    local rc = ffi.C.gethostname(buf, buf_size)
    if rc ~= 0 then
        error(('gethostname: %s'):format(errno.strerror()), 0)
    end
    return ffi.string(buf)
end

return {
    get = get,
}
```
</details>

### PR DESCRIPTION
**Pull request by Lord-KA**
Hostname is a useful piece of information in state reports. So it was decided to add it to box.info.

Hostname is obtained on requested and is not cached.

Closes #8605

@TarantoolBot document
Title: Add hostname to box.info

This patch adds hostname to box.info, it can be useful e.g. to supplement various instance state reports. It is not cached and is requested on each call.

### CODE REVIEW

File: changelogs/unreleased/gh-8605-add-gethostname.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## feature/box
+
+* Added hostname to box.info, it is obtained on request (gh-8605).
```
Comment (p7nov) [inline changelogs/unreleased/gh-8605-add-gethostname.md:3]:
```suggestion
* Added a new `box.info` parameter `hostname` (gh-8605).
```

Though I don't know if "It is obtained on request" is really important for a brief changelog entry. You can leave it if it is. Just make is a full separate sentence.

**Review by p7nov — APPROVED** [review]

File: src/box/lua/info.c [hunk 2]
```diff
@@ -650,6 +651,21 @@ lbox_schema_version(struct lua_State *L)
 	return 1;
 }
 
+/* gethostname() Lua interface inside box.info. */
+static int
+lbox_info_hostname(struct lua_State *L)
+{
+	char buffer[_SC_HOST_NAME_MAX + 1];
```
Comment (Totktonada) [inline src/box/lua/info.c:658]: It is not a value of max hostname length, it is ID of the corresponding parameter in sysconf.

```sh
$ cc --std=c89 -Wall -Wextra -x c <(echo -e '#include <stdio.h>\n#include<unistd.h>\n\nint main() { printf("_SC_HOST_NAME_MAX: %d\\n", _SC_HOST_NAME_MAX); return 0; }') && ./a.out; rm a.out

_SC_HOST_NAME_MAX: 180
```

/usr/include/bits/confname.h:

```c
/* Values for the NAME argument to `pathconf' and `fpathconf'.  */
enum
  {
    <...>
    _SC_HOST_NAME_MAX,
#define _SC_HOST_NAME_MAX               _SC_HOST_NAME_MAX
```

File: changelogs/unreleased/gh-8605-add-gethostname.md (hunk 1 above)
Comment (Lord-KA) [reply changelogs/unreleased/gh-8605-add-gethostname.md:3]: Agreed.

File: src/box/lua/info.c (hunk 2 above)
Comment (Lord-KA) [reply src/box/lua/info.c:658]: As expected, the mac has 255 default:
```
sysconf(_SC_HOST_NAME_MAX): 255
```

Comment (Lord-KA) [reply src/box/lua/info.c:658]: Same for FreeBSD:
```
sysconf(_SC_HOST_NAME_MAX): 255
```

File: changelogs/unreleased/gh-8605-add-gethostname.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## feature/box
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply changelogs/unreleased/gh-8605-add-gethostname.md:1]: Oh, I didn't recheck my source. Ok, removed it.

File: src/box/lua/info.c (hunk 2 above)
Comment (Lord-KA) [reply src/box/lua/info.c]: Added `enum` constant.

File: src/box/lua/info.c [hunk 4]
```diff
@@ -58,6 +59,18 @@
...
+ * Known upper limits for a hostname (without a zero-terminating
+ * byte):
+ *
+ * sysconf(_SC_HOST_NAME_MAX) == 64 on Linux.
+ * sysconf(_SC_HOST_NAME_MAX) == 255 on macOS.
+ * sysconf(_SC_HOST_NAME_MAX) == 255 on BSD.
+ *
+ * The constant value is used to simplify the code.
+ */
+enum { TT_HOST_NAME_MAX = 255 };
```
Comment (ligurio) [inline src/box/lua/info.c:72]: I would use a macro, not a constant number, because:

> Host names are limited to MAXHOSTNAMELEN (from <sys/param.h>) characters, currently 256. This includes the terminating NUL character. Note that the corresponding POSIX definition HOST_NAME_MAX in <limits.h> does [not](https://man.openbsd.org/sethostname#not) include the terminating NUL character. [^1]

[^1]: https://man.openbsd.org/sethostname

File: test/box/info.test.lua [hunk 5]
```diff
@@ -17,6 +17,8 @@ box.info.status
 string.len(box.info.uptime) > 0
 string.match(box.info.uptime, '^[1-9][0-9]*$') ~= nil
 box.info.cluster.uuid == box.space._schema:get{'cluster'}[2]
+type(box.info.hostname) == 'string'
```
Comment (ligurio) [inline test/box/info.test.lua:20]: After this `os.execute("hostname") ~= ''` to make sure that real hostname is not an empty string.

**Review by ligurio — COMMENTED** [review]
See my comments.

File: test/box/info.test.lua (hunk 5 above)
Comment (Lord-KA) [reply test/box/info.test.lua:20]: It seems like `os.execute()` doesn't work this way:
```
tarantool> os.execute('hostname') == 'gkashkin-vps'
gkashkin-vps
---
- false
...

tarantool> os.execute('hostname')
gkashkin-vps
---
- 0
...

tarantool> os.execute('hostname') == 0
gkashkin-vps
---
- true
...
```

Comment (Lord-KA) [reply test/box/info.test.lua:20]: (And I check that it is not empty string the line below.)

File: src/box/lua/info.c (hunk 4 above)
Comment (Totktonada) [reply src/box/lua/info.c:72]: There is no very clear definition in POSIX: is the `HOST_NAME_MAX` a constant/macro or a sysconf value? Why it is in the curly brackets? What header should define it?

Why `limits.h` on Linux actually defines the constant only if `__USE_POSIX` is defined?

Why `MAXHOSTNAMELEN` is defined on Linux (in `sys/param.h`) as `HOST_NAME_MAX` (without +1?), while macOS man page statement regarding `MAXHOSTNAMELEN` meaning is similar to what you quoted from OpenBSD (+1 is needed).

As for me, there are too many differences and nuances across OSes in those constants to play with them. We tried it with `LOG_MAKEPRI` (#8269) and lost.

`sysconf()` looks as a good way to obtain the value from the system, but we don't want complicate this code with dynamic allocations.

I bet 256 bytes buffer is good enough for the at least next 10 years.

File: src/box/lua/info.c [hunk 6]
```diff
@@ -650,6 +663,22 @@ lbox_schema_version(struct lua_State *L)
 	return 1;
 }
 
+/* gethostname() Lua interface inside box.info. */
```
Comment (Totktonada) [inline src/box/lua/info.c:666]: Nit: function/struct/field is usually described using a documentation-style comment: `/** <...> */`.

File: src/box/lua/info.c [hunk 7]
```diff
@@ -650,6 +663,22 @@ lbox_schema_version(struct lua_State *L)
...
+static int
+lbox_info_hostname(struct lua_State *L)
+{
+	char buffer[TT_HOST_NAME_MAX + 1];
+	int status = gethostname(buffer, TT_HOST_NAME_MAX + 1);
```
Comment (Totktonada) [inline src/box/lua/info.c:671]: Nit: Usually named `rc` across our codebase.

**Review by Totktonada — APPROVED** [review]

File: src/box/lua/info.c (hunk 6 above)
Comment (Lord-KA) [reply src/box/lua/info.c:666]: I would prefer to remove this comment entirely, as the nearby functions don't have one and the code is self-explanatory, but checkpatch is opposed to that. Otherwise, added extra `*`.

File: src/box/lua/info.c (hunk 4 above)
Comment (ligurio) [reply src/box/lua/info.c:72]: I suspect you address these question to me, not to Gleb.
I believe hardcoded constants are bad. In discussion with Gleb I suggest to use something like this:

```
#ifndef HOST_NAME_MAX
#define HOST_NAME_MAX 256
#endif
```

> sysconf() looks as a good way to obtain the value from the system, but we don't want complicate this code with dynamic allocations.

Probably it is good enough, but `sysconf` is not called in the patch.

File: src/box/lua/info.c (hunk 7 above)
Comment (olegrok) [inline src/box/lua/info.c:671]:
```suggestion
	int status = gethostname(buffer, sizeof(buffer));
```

nit: probably it looks a bit better.

File: src/box/lua/info.c (hunk 4 above)
Comment (Totktonada) [reply src/box/lua/info.c:72]: // Sorry for the long read. I started from a short note, but then found that not everything is cleanly explained. A bit vague topic.

TL;DR: I don't mind your proposal, but don't see benefits.

----

> I suspect you address these question to me, not to Gleb.

The questions are to the commitee behind POSIX and to libc implementations. They're here to illustrate difficulties with those constants.

Gleb asked me to summarize my opinion on using the hardcoded constant, so I did the attempt to illustrate my doubts about limits.h and sys/param.h constants.

> I believe hardcoded constants are bad.

As for me, the always-the-same constant is less error-prone comparing to the depend-on-os-and-build-flags constant.

At worst, when the world will be changed by future implemetations of future standards, ancient tarantool versions will not able to show insanely long host names. I don't feel it as a problem.

The behavior for the hardcoded constant can be described this way: till version X.Y.Z only hostnames up to 256 bytes long were shown.

At the opposite: say, we change a build host for packages and it appears somehow that `__USE_POSIX` becomes enabled and it may change tarantool's behavior, because the constant has nothing in common with the run host, which may have more recent `gethostname()` implementation with its own limits.

The behavior for the non-obviously-constructed constant... is hard to describe. It may be a problem for one who tries to reproduce a problem around.

Not a big problem, though. I wouldn't mind the proposal, just don't see benefits.

----

Regarding `sysconf()`.

We discussed this option with Gleb and agreed that it would add an unnecessary complication to the code. VLA is commonly considered unsafe (stack overflow -> denial of the service). Maybe sysconf's value is perfectly limited, but maybe there are some attack vectors. I'm not sure, so I would avoid it.

So, `sysconf()` means a dynamic allocation of a temporary buffer and, I guess, fiber region fits good here. A bit more code -- allocation and releasing the memory. Not so much, but...

I would like to avoid bikeshedding discussions around the best allocator: shouldn't we use cord_ibuf? Or, maybe, just use xmalloc, but move the allocation to the initialization code?

I mean, the value is limited in the real world, at least by [RFC 1025](https://www.rfc-editor.org/rfc/rfc1035#section-2.3.1) (section 2.3.1):

> There are also some restrictions on the length. Labels must be 63 characters or less.

macOS developers think that it should be 255. Okay, let's just pick up this limit.

It is anyway a constant, so the dynamic allocation here feels unnecessary.

---

BTW, there is no yield between the `gethostname()` and `lua_pushstring()` and all calls are from the TX threads. We can use `static` for the buffer.

I'll add it as a separate comment.

File: src/box/lua/info.c [hunk 8]
```diff
@@ -650,6 +663,22 @@ lbox_schema_version(struct lua_State *L)
 	return 1;
 }
 
+/* gethostname() Lua interface inside box.info. */
+static int
+lbox_info_hostname(struct lua_State *L)
+{
+	char buffer[TT_HOST_NAME_MAX + 1];
```
Comment (Totktonada) [inline src/box/lua/info.c:670]: There is no yield between the `gethostname()` and `lua_pushstring()` and all calls are from the TX threads. We can use `static` for the buffer.

File: src/box/lua/info.c (hunk 6 above)
Comment (Totktonada) [reply src/box/lua/info.c:666]: > added extra `*`.

Forgot to push?

File: src/box/lua/info.c (hunk 8 above)
Comment (olegrok) [reply src/box/lua/info.c:670]: There was similar issue when somebody tried to use static buffers in GC callback.

E.g. https://github.com/tarantool/tarantool/commit/c302360f0e27b6c1c30eb5787557bc5949db1336#diff-01cf72a8b25f913884f0b9d67ec342db4828a0d27490241f36e969db766f2f48R96

File: test/box/info.test.lua (hunk 5 above)
Comment (ligurio) [reply test/box/info.test.lua:20]: Sorry, I mean `io.popen` or `popen.shell`. Execute `hostname` via popen, obtain value via stdout and compare it with value obtained using `box.info.hostname`.

File: src/box/lua/info.c (hunk 8 above)
Comment (Totktonada) [reply src/box/lua/info.c:670]: Ouch, yes. If a GC callback invokes `box.info.hostname`, we rewrite the buffer before `lua_pushstring()` catches it.

It would be a strange GC callback, but possible. Okay, let's allocate it on the stack on each call.

File: src/box/lua/info.c (hunk 4 above)
Comment (ligurio) [reply src/box/lua/info.c:72]: Discussed with  @Totktonada verbally and decided to left it as is.

Comment (Totktonada) [reply src/box/lua/info.c:72]: > BTW, there is no yield between the `gethostname()` and `lua_pushstring()` and all calls are from the TX threads. We can use static for the buffer.

Discussed and, as Oleg correctly said, it is wrong due to possible GC callback invocation.

File: test/box/info.test.lua (hunk 5 above)
Comment (Lord-KA) [reply test/box/info.test.lua:20]: Sure, will do.

File: src/box/lua/info.c (hunk 7 above)
Comment (Lord-KA) [reply src/box/lua/info.c:671]: Renamed to `rc`.

Comment (Lord-KA) [reply src/box/lua/info.c:671]: Good idea, thanks.

File: test/box/info.test.lua (hunk 5 above)
Comment (Lord-KA) [reply test/box/info.test.lua:20]: Added [another test](https://github.com/tarantool/tarantool/blob/ce73fa2f5697f2017b45730434b10bbf34d11aa3/test/box-luatest/gh_8605_add_box_info_hostname_test.lua) for that.

Comment (ligurio) [reply test/box/info.test.lua:20]: Thanks!

**Review by ligurio — APPROVED** [review]
LGTM

**Review by ligurio — APPROVED** [review]
LGTM

File: test/box-luatest/gh_8605_add_box_info_hostname_test.lua [hunk 9]
```diff
@@ -0,0 +1,18 @@
+local popen = require('popen')
+local t = require('luatest')
+
+local g = t.group()
+
+g.test_hostname_result = function()
+    local ph = popen.shell("hostname", 'r')
```
Comment (Totktonada) [inline test/box-luatest/gh_8605_add_box_info_hostname_test.lua:7]: Nit: `ph:close()` is desirable. It is collected by GC anyway, but the GC delay at freeing a tightly limited resource (such as a file descriptor) may cause unexpected effects under certain workloads. There may be a peak, where it hits the file descriptor limit.

Unlikely we'll be hit by this problem there, just a suggestion regarding the good practice.

File: changelogs/unreleased/gh-8605-add-gethostname.md (hunk 3 above)
Comment (Totktonada) [inline changelogs/unreleased/gh-8605-add-gethostname.md:1]: I'm ready to push, but there is a conflict. Please, rebase on a fresh master and set the full-ci label.

File: test/box-luatest/gh_8605_add_box_info_hostname_test.lua (hunk 9 above)
Comment (Lord-KA) [reply test/box-luatest/gh_8605_add_box_info_hostname_test.lua:7]: My bad, missed `close()`.

Comment (Lord-KA) [reply test/box-luatest/gh_8605_add_box_info_hostname_test.lua:7]: Fixed.

File: changelogs/unreleased/gh-8605-add-gethostname.md (hunk 3 above)
Comment (Lord-KA) [reply changelogs/unreleased/gh-8605-add-gethostname.md:1]: Done.

File: test/box/info.test.lua (hunk 5 above)
Comment (Lord-KA) [reply test/box/info.test.lua:20]: Well, there is a problem with hostname on [redos](https://github.com/tarantool/tarantool/actions/runs/5030912165/jobs/9023648604?pr=8651):
```
[036] sh: hostname: command not found
[036] #   expected: {exit_code = 0, state = "exited"}
[036] #   actual: {exit_code = 127, state = "exited"}
```

Comment (Totktonada) [reply test/box/info.test.lua:20]: I suspect there is no hostname command there. Let's skip the test if it is so (either by the exit code or use 'type/command/which hostname' beforehand).

File: changelogs/unreleased/gh-8605-add-gethostname.md (hunk 3 above)
Comment (Totktonada) [reply changelogs/unreleased/gh-8605-add-gethostname.md:1]: Waiting for the test fix.

File: test/box/info.test.lua (hunk 5 above)
Comment (Totktonada) [reply test/box/info.test.lua:20]: The exit code 127 is [standartized](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_08_02).

Comment (Lord-KA) [reply test/box/info.test.lua:20]: Added `skip_if()` to the test.

File: changelogs/unreleased/gh-8605-add-gethostname.md (hunk 3 above)
Comment (Lord-KA) [reply changelogs/unreleased/gh-8605-add-gethostname.md:1]: @Totktonada, the test is fixed, PR ready for merge.

File: test/box/info.test.lua (hunk 5 above)
Comment (Totktonada) [reply test/box/info.test.lua:20]: No `:close()` at skipping the test, but this is negligible due to popen object's GC hook.
